### ISSUE TITLE
Poor performance receiving large message

### ISSUE THREAD

**Issue by kohtala** [amqp/rhea#381]
A node:16.13.2 process was receiving messages of about 50MB in size. The same process implements a REST API interface that shows over 5 s latencies even in simplest responses during the message reception.

Profiling shows a lot of time is spent in garbage collector.

Looking into it further, I find `Connection.prototype.input` receives the input in 64KB buffers and collects the inputs using `Buffer.concat` until the resulting buffer contains a complete frame. The sender sends the whole 50MB in one frame. Should the sender use smaller frames, `Incoming.prototype.on_transfer` would use the same strategy of `Buffer.concat` to combine the payload out of the frames frame by frame.

I believe the about 762 consecutive allocations of x*64KB sizes buffer hits the garbage collector hard stalling the whole process.

**Comment by kohtala** [comment]
The `Connection.prototype.input` would know the size to allocate from the first four bytes carrying the size of the frame. Thus by some code rearrangement it would be possible to have the whole buffer allocated early and later filled with `Buffer.copy`. That would be the most efficient way to handle that I could figure out.

For the `Incoming.prototype.on_transfer` the complete size is not so easily found. But that could be made better by cumulating the message data into a list that could then be concatenated with single `Buffer.concat`. That should be easier on the garbage collector.

I need this fixed, so I'll try to see if I can come up with something. If anyone notices this and has ideas before that, please share.

**Comment by grs** [comment]
Does setting max_frame_size on the receiving connection resolve the issue for you?

**Comment by kohtala** [comment]
I think `max_frame_size` can make it better by staging the `Buffer.concat` calls such that there is less of very large allocation. With 50MB frame, towards the end there is a long series of increasingly large allocations. It'll cumulate in smaller allocations before making the less frequent large allocation in `Incoming.prototype.on_transfer`.

I see the REST API latency grow already early in the transfer. So even the "smaller" frame already has some impact I'd rather see gone.

We do not really have a limit of 50MB. I worry a bit that `max_frame_size` alone there will be another bug hunt with a larger message at some point eating any savings in avoiding a real fix now. That could be more costly than spending some time thinking of a solution now.

The `Transport.prototype.read` now returns only the offset. I wonder if a change of that interface would be breaking change requiring a new major. The `Transport.read` is on the typings although it seems like an internal method. It needs some way of conveying back the size of the incomplete frame. Maybe even the new Buffer for remaining bytes plus what is needed for the complete frame.

**Comment by kohtala** [comment]
I measured the performance impact of 2MB `max_frame_size`. Garbage collection is reduced to about 1/10.

**Comment by kohtala** [comment]
I did not find simple and reliable fix for the buffer handling at reception. It started getting too complex. So I wrote only a partial fix to handle the large allocations at receiving transfer frames.

I tried testing it adding a 50MB test case into the fragmentation tests. I noticed it did not really fragment messages it was sending. I fixed that.

After having that fixed, there is a problem that the sender will not send any frames if the recipient don't have window open to receive them all. So sending stalls there because 50MB divided into 16KB frames exceeds 2048 frames available in the window. I wrote something that could fix it, but then the receiver did not give any more transfer window back to sender once it had received the 2048 frames, so it got stuck there.

Now I have to take a break. This is taking more effort than I can put into this now. I added the sending fix to the PR, but I have not tested if it really works as intended. It did not seem to break existing tests.
